### PR TITLE
Address SQLite handling of na

### LINKED ISSUE #103: Ingest fails for certain (older?) versions of sqlite

### PR DESCRIPTION
**Pull request by shntnu**
Fixes #103 

https://github.com/cytomining/cytominer-database/pull/104#issuecomment-511440383 explains why this worked without needing to change any code

### CODE REVIEW

File: tests/data_b/B01-2/Cells.csv [hunk 1]
```diff
@@ -3,7 +3,7 @@ ImageNumber,ObjectNumber,AreaShape_Area,AreaShape_Center_X,AreaShape_Center_Y,Ar
...
...32,0.59834892009626228,0.4143413817510756,0.77025345622119823,5.692387364060072,5.9840584523414151,5.4965437788018452,5.1341273951320545,5.6051145798737947,4.8497695852534566,4.2982910409114448,4.3586848223181667,4.2292626728110605,5.4189539098912478,5.4360677515775482,5.3723502304147468,6.3412739513205594,6.6542676851544336,6.1232718894009226,2.3252167598501128,2.3108514018566204,2.2336451619833588,2.0715687492735713,2.2260161345470069,1.8740236953150555,1.9986916779247255,1.9813685528367178,1.9177008971728069,2.1885663549705163,2.1449626577399474,2.1836805877203109,2.4173304139314777,2.3886959349397232,2.4027653772711073,10.241469753625935,8.2468398557663196,10.960587132876046,14.619503368545208,10.531876863324458,15.787569071333008,6.379173671897985,4.8696846203574831,7.2656413918324878,6.4987396617804141,5.1353181807302155,7.4581294994584715,11.823459588855387,9.8190277848707304,12.720979156490898,2.7294552399577237,2.5691748433536246,2.7901324619337848,4.2561972328073594,4.4600258478791535,4.1061137420629015,1.3569921081023966,0.81074629381187391,1.7038939030346791,1.3645991926004373,0.7841320450057967,1.7702019579944359,3.3595695728420245,3.4671681527303688,3.3387230032491666
```
Comment (shntnu) [inline tests/data_b/B01-2/Cells.csv]: I changed the 6th column of this row to `na`

**Comment by shntnu** [comment]
Fails on AppVeyor https://ci.appveyor.com/project/shntnu/cytominer-database/build/1.0.203/job/l0me0e7gl4axo88g

    '(exceptions.ValueError) could not convert string to float: na'


but passes on Travis

**Comment by gwaybio** [comment]
interesting - this fails now with an error:

looks like we will need to configure `pd.read_sql()` a bit more. 

@shntnu @bethac07 is `na` typically how a missing value will be coded in the `.csv` files?

```
tests/test_ingest.py:37: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../../virtualenv/python2.7.14/lib/python2.7/site-packages/pandas/io/sql.py:392: in read_sql
    parse_dates=parse_dates, columns=columns, chunksize=chunksize)
../../../virtualenv/python2.7.14/lib/python2.7/site-packages/pandas/io/sql.py:1039: in read_table
    chunksize=chunksize)
../../../virtualenv/python2.7.14/lib/python2.7/site-packages/pandas/io/sql.py:731: in read
    self._harmonize_columns(parse_dates=parse_dates)
../../../virtualenv/python2.7.14/lib/python2.7/site-packages/pandas/io/sql.py:853: in _harmonize_columns
    self.frame[col_name] = df_col.astype(col_type, copy=False)
../../../virtualenv/python2.7.14/lib/python2.7/site-packages/pandas/core/generic.py:5691: in astype
    **kwargs)
../../../virtualenv/python2.7.14/lib/python2.7/site-packages/pandas/core/internals/managers.py:531: in astype
    return self.apply('astype', dtype=dtype, **kwargs)
../../../virtualenv/python2.7.14/lib/python2.7/site-packages/pandas/core/internals/managers.py:395: in apply
    applied = getattr(b, f)(**kwargs)
../../../virtualenv/python2.7.14/lib/python2.7/site-packages/pandas/core/internals/blocks.py:534: in astype
    **kwargs)
../../../virtualenv/python2.7.14/lib/python2.7/site-packages/pandas/core/internals/blocks.py:633: in _astype
    values = astype_nansafe(values.ravel(), dtype, copy=True)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
arr = array([1.0536865809980278, 1.1503667249090357, 1.0698111464164657,
       1.03... 1.108436713089547, 1.395106670685756, 1.2661806217001421],
      dtype=object)
dtype = dtype('float64'), copy = True, skipna = False
    def astype_nansafe(arr, dtype, copy=True, skipna=False):
        """
        Cast the elements of an array to a given dtype a nan-safe manner.
    
        Parameters
        ----------
        arr : ndarray
        dtype : np.dtype
        copy : bool, default True
            If False, a view will be attempted but may fail, if
            e.g. the item sizes don't align.
        skipna: bool, default False
            Whether or not we should skip NaN when casting as a string-type.
    
        Raises
        ------
        ValueError
            The dtype was a datetime64/timedelta64 dtype, but it had no unit.
        """
    
        # dispatch on extension dtype if needed
        if is_extension_array_dtype(dtype):
            return dtype.construct_array_type()._from_sequence(
                arr, dtype=dtype, copy=copy)
    
        if not isinstance(dtype, np.dtype):
            dtype = pandas_dtype(dtype)
    
        if issubclass(dtype.type, text_type):
            # in Py3 that's str, in Py2 that's unicode
            return lib.astype_unicode(arr.ravel(),
                                      skipna=skipna).reshape(arr.shape)
    
        elif issubclass(dtype.type, string_types):
            return lib.astype_str(arr.ravel(),
                                  skipna=skipna).reshape(arr.shape)
    
        elif is_datetime64_dtype(arr):
            if is_object_dtype(dtype):
                return tslib.ints_to_pydatetime(arr.view(np.int64))
            elif dtype == np.int64:
                return arr.view(dtype)
    
            # allow frequency conversions
            if dtype.kind == 'M':
                return arr.astype(dtype)
    
            raise TypeError("cannot astype a datetimelike from [{from_dtype}] "
                            "to [{to_dtype}]".format(from_dtype=arr.dtype,
                                                     to_dtype=dtype))
    
        elif is_timedelta64_dtype(arr):
            if is_object_dtype(dtype):
                return tslibs.ints_to_pytimedelta(arr.view(np.int64))
            elif dtype == np.int64:
                return arr.view(dtype)
    
            # in py3, timedelta64[ns] are int64
            if ((PY3 and dtype not in [_INT64_DTYPE, _TD_DTYPE]) or
                    (not PY3 and dtype != _TD_DTYPE)):
    
                # allow frequency conversions
                # we return a float here!
                if dtype.kind == 'm':
                    mask = isna(arr)
                    result = arr.astype(dtype).astype(np.float64)
                    result[mask] = np.nan
                    return result
            elif dtype == _TD_DTYPE:
                return arr.astype(_TD_DTYPE, copy=copy)
    
            raise TypeError("cannot astype a timedelta from [{from_dtype}] "
                            "to [{to_dtype}]".format(from_dtype=arr.dtype,
                                                     to_dtype=dtype))
    
        elif (np.issubdtype(arr.dtype, np.floating) and
              np.issubdtype(dtype, np.integer)):
    
            if not np.isfinite(arr).all():
                raise ValueError('Cannot convert non-finite values (NA or inf) to '
                                 'integer')
    
        elif is_object_dtype(arr):
    
            # work around NumPy brokenness, #1987
            if np.issubdtype(dtype.type, np.integer):
                return lib.astype_intsafe(arr.ravel(), dtype).reshape(arr.shape)
    
            # if we have a datetime/timedelta array of objects
            # then coerce to a proper dtype and recall astype_nansafe
    
            elif is_datetime64_dtype(dtype):
                from pandas import to_datetime
                return astype_nansafe(to_datetime(arr).values, dtype, copy=copy)
            elif is_timedelta64_dtype(dtype):
                from pandas import to_timedelta
                return astype_nansafe(to_timedelta(arr).values, dtype, copy=copy)
    
        if dtype.name in ("datetime64", "timedelta64"):
            msg = ("The '{dtype}' dtype has no unit. "
                   "Please pass in '{dtype}[ns]' instead.")
            raise ValueError(msg.format(dtype=dtype.name))
    
        if copy or is_object_dtype(arr) or is_object_dtype(dtype):
            # Explicit copy, or required since NumPy can't view from / to object.
>           return arr.astype(dtype, copy=True)
E           ValueError: could not convert string to float: na
../../../virtualenv/python2.7.14/lib/python2.7/site-packages/pandas/core/dtypes/cast.py:702: ValueError
```

**Comment by bethac07** [comment]
It should be `NaN`

**Comment by bethac07** [comment]
Actually, appears to be `nan` in CP 3+ 
image

**Comment by gwaybio** [comment]
great! Thanks Beth!

**Comment by gwaybio** [comment]
I changed `na` to `nan` in 70184d383dca649d6259b9c385267ac180a239e3 using:

```python
file = "tests/data_b/B01-2/Cells.csv"
df = pd.read_csv(file)
df.loc[4, "AreaShape_Compactness"] = "nan"
df.to_csv(file, sep=',', index=False)
```

This seems to have worked, although I am not quite sure why (haven't searched much yet). Note that the change in 70184d383dca649d6259b9c385267ac180a239e3 seems to have updated the amount of significant digits in some features, which is why it seems like the entire `csv` file was updated.

**Comment by shntnu** [comment]
> Note that the change in [70184d3](https://github.com/cytomining/cytominer-database/commit/70184d383dca649d6259b9c385267ac180a239e3) seems to have updated the amount of significant digits in some features, which is why it seems like the entire `csv` file was updated.

I reverted the change, then remade it by hand, to reduce diffs. Hope that's ok @gwaygenomics

**Comment by shntnu** [comment]
> This seems to have worked, although I am not quite sure why (haven't searched much yet)

IIRC it worked ok on most machines. 

Also this: https://github.com/cytomining/cytominer-database/pull/104#issuecomment-391873764

So this might be a tough to reproduce. I verified this worked fine on sqlite v3.23.1 on OSX and 3.8.2 on linux 3.13.0-151-generic. It passes on Travis. I'd say that's good enough, given that this error is hard to reproduce.

**Comment by gwaybio** [comment]
> I reverted the change, then remade it by hand, to reduce diffs. Hope that's ok @gwaygenomics

Sounds great! Thanks @shntnu

**Comment by gwaybio** [comment]
> This seems to have worked, although I am not quite sure why (haven't searched much yet)

I think it is just that pandas knows to convert `nan` to a missing value, but not `na`. Since cellprofiler outputs `nan` I think this is good to merge!

**Review by gwaybio — APPROVED** [review]

**Comment by shntnu** [comment]
I have a hunch this https://stackoverflow.com/questions/15569745/store-nan-values-in-sqlite-database will become relevant at some point so adding that reference here

**Pull request merged**